pkg/controller/deployment_sync.go

```diff
@@ -441,56 +445,39 @@ func (dc *controller) scale(ctx context.Context, deployment *v1alpha1.MachineDep
+		klog.V(3).Infof("TestLog: machineDeployment: %s , replicasToAdd: %d, maxAllowedSize: %d, allMachineSetReplicas: %d", deployment.Name, allowedSize, allISsReplicas)
 		deploymentReplicasToAdd := allowedSize - allISsReplicas
 
-		// The additional replicas should be distributed proportionally amongst the active
-		// machine sets from the larger to the smaller in size machine set. Scaling direction
-		// drives what happens in case we are trying to scale machine sets of the same size.
-		// In such a case when scaling up, we should scale up newer machine sets first, and
-		// when scaling down, we should scale down older machine sets first.
+		// During scale-down, the additional replicas should be distributed proportionally amongst the active
+		// machine sets from the larger to the smaller in size machine set.
+		// We should scale down older machine sets first if machine sets are of equal size.
+
 		var scalingOperation string
+		nameToSize := make(map[string]int32)
+		deploymentReplicasAdded := int32(0)
 		switch {
 		case deploymentReplicasToAdd > 0:
-			sort.Sort(MachineSetsBySizeNewer(allISs))
 			scalingOperation = "up"
-
+			nameToSize = dc.scaleNewMachineSet(newIS, allISs, deploymentReplicasToAdd, deployment)
+			deploymentReplicasAdded = deploymentReplicasToAdd
 		case deploymentReplicasToAdd < 0:
-			sort.Sort(MachineSetsBySizeOlder(allISs))
 			scalingOperation = "down"
+			sort.Sort(MachineSetsBySizeOlder(allISs))
+			nameToSize, deploymentReplicasAdded = dc.scaleMachineSetsProportionally(allISs, deploymentReplicasToAdd, deployment)
 		}
 
-		// Iterate over all active machine sets and estimate proportions for each of them.
-		// The absolute value of deploymentReplicasAdded should never exceed the absolute
-		// value of deploymentReplicasToAdd.
-		deploymentReplicasAdded := int32(0)
-		nameToSize := make(map[string]int32)
-		for i := range allISs {
-			is := allISs[i]
-
-			// Estimate proportions if we have replicas to add, otherwise simply populate
-			// nameToSize with the current sizes for each machine set.
-			if deploymentReplicasToAdd != 0 {
-				proportion := GetProportion(is, *deployment, deploymentReplicasToAdd, deploymentReplicasAdded)
-
-				nameToSize[is.Name] = (is.Spec.Replicas) + proportion
-				deploymentReplicasAdded += proportion
-			} else {
-				nameToSize[is.Name] = (is.Spec.Replicas)
-			}
-		}
-
-		// Update all machine sets
 		for i := range allISs {
 			is := allISs[i]
 
-			// Add/remove any leftovers to the largest machine set.
+			// Incorporate any leftovers to the largest machine set.
 			if i == 0 && deploymentReplicasToAdd != 0 {
 				leftover := deploymentReplicasToAdd - deploymentReplicasAdded
 				nameToSize[is.Name] = nameToSize[is.Name] + leftover
 				if nameToSize[is.Name] < 0 {
 					nameToSize[is.Name] = 0
 				}
+				klog.V(3).Infof("TestLog: leftover proportion increase of %d done in largest machineSet %s", leftover, is.Name)
```

Comment:
Can you remove the word TestLog or remove this log if this is only required for testing?